content/docs/guides/driver-design.md

```diff
@@ -0,0 +1,223 @@
+and the second being the inline composite literal slice declaration `[]byte{x, x}`. Usually heap allocators have limitations on size allocated so it is likely that more than 4 bytes are allocated for every call.
+
+#### Heap allocations: Mitigations
+Luckily it can be relatively easy to mitigate heap allocations and eliminate them altogether. To do so with a driver that uses slices for bus communications one can include the array memory within the device struct:
+
+```go
+type Device struct {
+    spi      drivers.SPI
+    bufRead  [2]byte
+    bufWrite [2]byte
+}
+
+func (d *Device) readRegister(addr uint8) (value uint8, err error) {
+    d.bufWrite = [2]byte{addr, 0}
+    err= d.spi.Tx(d.bufWrite[:], d.bufRead[:])
+    return d.bufRead[1], err
+}
+
+
+func (d *Device) writeRegister(addr, newValue uint8) error {
+    d.bufWrite = [2]byte{addr, newValue}
+    return d.spi.Tx(d.bufWrite[:], d.bufRead[:])
+}
+```
+
+This can solve many such cases, but there are few cases where data to be read/written is also variable length.
+The semantics for reading and writing data over the SPI bus are hypothetical, what is important is to take note on how to 
+deal with variable length buffers.
+```go
+const maxCommSize = 32 // Never send more than 32 bytes in a single transaction.
+
+type Device struct {
+    spi   drivers.SPI
+    wdata [maxCommSize]byte
+    rdata [maxCommSize]byte
+}
+
+func (d *Device) Configure() error {
+    // Since the argument buffer is not passed into Tx call,
+    // data does NOT escape and may be stack allocated.
+    data := []byte{1, 2, 3, 4}
+    return d.writeData(CONFIG_REG, data)
+}
+
+func (d *Device) writeData(addr uint8, data []byte) error {
+    d.data[0] = addr
+    n := copy(d.data[1:], data)
+    return d.spi.Tx(d.wdata[:n+1], d.rdata[:n+1])
+}
+
+func (d *Device) readData(addr uint8, data []byte) error {
+    for i := range d.wdata {
+        d.wdata[i] = 0 // Clear write data buffer.
+    }
+    dlen := len(data)
+    d.wdata[0] = addr
+    err := d.spi.Tx(d.wdata[:dlen+1], d.rdata[:dlen+1])
+    if err != nil {
+        return err
+    }
+    copy(data, d.rdata[1:])
+    return nil
+}
+```
+
+### Multi-protocol devices
+Some devices support SPI and I2C interfaces (UART, RS485, etc.). What's the ideal approach to this? Should one generate a separate driver for each protocol?
+
+The answer is, as usual in engineering "it depends". That said more often than not a peripheral which supports different protocols shares a whole lot of 
+functioning. Take for example the Honeywell HSC TruStability pressure sensors. They are a VERY simple pressure sensor which come in SPI and I2C varieties.
+One can write a driver that shares the underlying shared logic like so:
+
+```go
+package hsc
+
+// device implements the shared logic between Honeywell HSC pressure sensor varieties.
+type device struct {
+    // calibration factors.
+    k1, k2, k3 uint32
+    lastRead   uint32
```

Comment:
A piece of advice (to add above, in the beginning of the guide) to separate "update" of value from "reading" value, client being responsible to poll Update regularly?